In [1]:
import pandas as pd
from sklearn.metrics import pairwise_distances
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
book_data=pd.read_csv('book.csv',encoding='latin-1')
book_data

Unnamed: 0  User.ID                                         Book.Title  \
0              1   276726                                Classical Mythology   
1              2   276729                                       Clara Callan   
2              3   276729                               Decision in Normandy   
3              4   276736  Flu: The Story of the Great Influenza Pandemic...   
4              5   276737                             The Mummies of Urumchi   
...          ...      ...                                                ...   
9995        9996   162121       American Fried: Adventures of a Happy Eater.   
9996        9997   162121                              Cannibal In Manhattan   
9997        9998   162121                    How to Flirt: A Practical Guide   
9998        9999   162121                                           Twilight   
9999       10000   162129                       Kids Say the Darndest Things   

      Book.Rating  
0               5  
1               3  
2               6  
3               8  
4               6  
...           ...  
9995            7  
9996            9  
9997            7  
9998            8  
9999            6  

[10000 rows x 4 columns]

### EDA

In [3]:
book_data.shape

(10000, 4)

In [4]:
book_data.isnull().sum()

Unnamed: 0     0
User.ID        0
Book.Title     0
Book.Rating    0
dtype: int64

In [5]:
## there is no null entries

In [6]:
book_data.describe(include='all')

Unnamed: 0        User.ID      Book.Title  Book.Rating
count   10000.00000   10000.000000           10000  10000.00000
unique          NaN            NaN            9659          NaN
top             NaN            NaN  Fahrenheit 451          NaN
freq            NaN            NaN               5          NaN
mean     5000.50000   95321.249800             NaN      7.56630
std      2886.89568  117645.703609             NaN      1.82152
min         1.00000       8.000000             NaN      1.00000
25%      2500.75000    2103.000000             NaN      7.00000
50%      5000.50000    3757.000000             NaN      8.00000
75%      7500.25000  162052.000000             NaN      9.00000
max     10000.00000  278854.000000             NaN     10.00000

### Fahrenheit 451	is the most picked book with frequency 5

In [7]:
book_data.dtypes

Unnamed: 0      int64
User.ID         int64
Book.Title     object
Book.Rating     int64
dtype: object

In [8]:
book_data['User.ID'].nunique()

2182

In [9]:
book_data['Book.Title'].nunique()

9659

In [10]:
#there are 9659 unique book in the list

In [11]:
book_data['Book.Title'].unique()

array(['Classical Mythology', 'Clara Callan', 'Decision in Normandy', ...,
       'How to Flirt: A Practical Guide', 'Twilight',
       'Kids Say the Darndest Things'], dtype=object)

In [12]:
book_data.value_counts()

Unnamed: 0  User.ID  Book.Title                                                                Book.Rating
1           276726   Classical Mythology                                                       5              1
6671        3416     At Wits End                                                               1              1
6664        3408     Intimate Evil: A Novel                                                    8              1
6665        3416     Magic Goes Away                                                           8              1
6666        3416     Beloved Exile                                                             8              1
                                                                                                             ..
3334        805      Kissing the Witch: Old Tales in New Skins                                 7              1
3335        805      The DREAMS OUR STUFF IS MADE OF: How Science Fiction Conquered the World  4             

In [13]:
book_data.rename(columns={'User.ID':'User_ID'},inplace=True)
book_data

Unnamed: 0  User_ID                                         Book.Title  \
0              1   276726                                Classical Mythology   
1              2   276729                                       Clara Callan   
2              3   276729                               Decision in Normandy   
3              4   276736  Flu: The Story of the Great Influenza Pandemic...   
4              5   276737                             The Mummies of Urumchi   
...          ...      ...                                                ...   
9995        9996   162121       American Fried: Adventures of a Happy Eater.   
9996        9997   162121                              Cannibal In Manhattan   
9997        9998   162121                    How to Flirt: A Practical Guide   
9998        9999   162121                                           Twilight   
9999       10000   162129                       Kids Say the Darndest Things   

      Book.Rating  
0               5  
1               3  
2               6  
3               8  
4               6  
...           ...  
9995            7  
9996            9  
9997            7  
9998            8  
9999            6  

[10000 rows x 4 columns]

### Collaborative Filtering-UBCF(cosine similarity score)

## Model Building

In [14]:
ubcf=pd.pivot_table(data=book_data,values='Book.Rating',index='User_ID',columns='Book.Title').fillna(0)
ubcf.index=book_data.User_ID.unique()
ubcf

Book.Title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
276726                       0.0                                         0.0   
276729                       0.0                                         0.0   
276736                       0.0                                         0.0   
276737                       0.0                                         0.0   
276744                       0.0                                         0.0   
...                          ...                                         ...   
162107                       0.0                                         0.0   
162109                       0.0                                         0.0   
162113                       0.0                                         0.0   
162121                       0.0                                         0.0   
162129                       0.0                                         0.0   

Book.Title   Repairing PC Drives &amp  '48  \
276726                            0.0  0.0   
276729                            0.0  0.0   
276736                            0.0  0.0   
276737                            0.0  0.0   
276744                            0.0  0.0   
...                               ...  ...   
162107                            0.0  0.0   
162109                            0.0  0.0   
162113                            0.0  0.0   
162121                            0.0  0.0   
162129                            0.0  0.0   

Book.Title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
276726                                                    0.0                       
276729                                                    0.0                       
276736                                                    0.0                       
276737                                                    0.0                       
276744                                                    0.0                       
...                                                       ...                       
162107                                                    0.0                       
162109                                                    0.0                       
162113                                                    0.0                       
162121                                                    0.0                       
162129                                                    0.0                       

Book.Title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
276726                                                    0.0              
276729                                                    0.0              
276736                                                    0.0              
276737                                                    0.0              
276744                                                    0.0              
...                                                       ...              
162107                                                    0.0              
162109                                                    0.0              
162113                                                    0.0              
162121                                                    0.0              
162129                                                    0.0              

Book.Title  01-01-00: A Novel of the Millennium  \
276726                                      0.0   
276729                                      0.0   
276736                                      0.0   
276737                                      0.0   
276744                                      0.0   
...                                         ...   
162107                                      0.0   
162109                                      0.0   
162113                                      0.0   
162121                                    

In [15]:
cosine_distances=1-pairwise_distances(X=ubcf,metric='cosine')
cosine_df=pd.DataFrame(data=cosine_distances)
cosine_df.index=book_data.User_ID.unique()
cosine_df.columns=book_data.User_ID.unique()
cosine_df



276726  276729  276736  276737  276744  276745  276747  276748  \
276726     1.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     1.0     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0     1.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
162107     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
162109     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
162113     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
162121     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
162129     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

        276751  276754  ...  162085  162091  162092  162095  162103  162107  \
276726     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
162107     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     1.0   
162109     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
162113     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
162121     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
162129     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

        162109  162113  162121  162129  
276726     0.0     0.0     0.0     0.0  
276729     0.0     0.0     0.0     0.0  
276736     0.0     0.0     0.0     0.0  
276737     0.0     0.0     0.0     0.0  
276744     0.0     0.0     0.0     0.0  
...        ...     ...     ...     ...  
162107     0.0     0.0     0.0     0.0  
162109     1.0     0.0     0.0     0.0  
162113     0.0     1.0     0.0     0.0  
162121     0.0     0.0     1.0     0.0  
162129     0.0     0.0     0.0     1.0  

[2182 rows x 2182 columns]

In [16]:
first_500_users=cosine_df.iloc[:500,:500]
first_500_users

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     1.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     1.0     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0     1.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
278390     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278396     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278398     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278401     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278407     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

        276751  276754  ...  278364  278376  278378  278388  278389  278390  \
276726     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
278390     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     1.0   
278396     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278398     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278401     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278407     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

        278396  278398  278401  278407  
276726     0.0     0.0     0.0     0.0  
276729     0.0     0.0     0.0     0.0  
276736     0.0     0.0     0.0     0.0  
276737     0.0     0.0     0.0     0.0  
276744     0.0     0.0     0.0     0.0  
...        ...     ...     ...     ...  
278390     0.0     0.0     0.0     0.0  
278396     1.0     0.0     0.0     0.0  
278398     0.0     1.0     0.0     0.0  
278401     0.0     0.0     1.0     0.0  
278407     0.0     0.0     0.0     1.0  

[500 rows x 500 columns]

In [17]:
np.fill_diagonal(a=first_500_users.to_numpy(),val=0)

In [18]:
first_500_users

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...     ...     ...     ...     ...     ...     ...   
278390     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278396     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278398     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278401     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
278407     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

        276751  276754  ...  278364  278376  278378  278388  278389  278390  \
276726     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
...        ...     ...  ...     ...     ...     ...     ...     ...     ...   
278390     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278396     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278398     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278401     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
278407     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

        278396  278398  278401  278407  
276726     0.0     0.0     0.0     0.0  
276729     0.0     0.0     0.0     0.0  
276736     0.0     0.0     0.0     0.0  
276737     0.0     0.0     0.0     0.0  
276744     0.0     0.0     0.0     0.0  
...        ...     ...     ...     ...  
278390     0.0     0.0     0.0     0.0  
278396     0.0     0.0     0.0     0.0  
278398     0.0     0.0     0.0     0.0  
278401     0.0     0.0     0.0     0.0  
278407     0.0     0.0     0.0     0.0  

[500 rows x 500 columns]

In [19]:
first_500_users.idxmax()

276726    276726
276729    276726
276736    276726
276737    276726
276744    276726
           ...  
278390    276726
278396    276726
278398    277988
278401    278246
278407    276726
Length: 500, dtype: int64

### Extracting books read by diffrent users

In [22]:
book_data[(book_data['User_ID']==278398) | (book_data['User_ID']==277988)]

Unnamed: 0  User_ID                                         Book.Title  \
1410        1411   277988                                           Shopgirl   
1899        1900   278398  Managing Technology and Innovation for Competi...   

      Book.Rating  
1410           10  
1899            8

In [23]:
book_data[(book_data['User_ID']==278407) | (book_data['User_ID']==276726)]

Unnamed: 0  User_ID                             Book.Title  Book.Rating
0              1   276726                    Classical Mythology            5
1901        1902   278407  The Unofficial Guide to Having a Baby           10

In [24]:
book_data[(book_data['User_ID']==276729) | (book_data['User_ID']==276726)]

Unnamed: 0  User_ID            Book.Title  Book.Rating
0           1   276726   Classical Mythology            5
1           2   276729          Clara Callan            3
2           3   276729  Decision in Normandy            6

In [25]:
book_data[(book_data['User_ID']==278390) | (book_data['User_ID']==276726)]

Unnamed: 0  User_ID                                         Book.Title  \
0              1   276726                                Classical Mythology   
1886        1887   278390  Technomanifestos: Visions from the Information...   
1887        1888   278390         The Wicked Priest: An Allegory in Two Acts   
1888        1889   278390                       Doctor Sax: Faust Part Three   
1889        1890   278390            Dorien's Diary: My Intimate Confessions   
1890        1891   278390                           Fg on Our Immigrant Ance   
1891        1892   278390                            Fg Cooks 3 Ancient Cuis   
1892        1893   278390                                  Fg Cooks American   
1893        1894   278390  The Gift of Motherhood : 10 Truths for Every M...   
1894        1895   278390  Chicken Soup for the Parent's Soul : 101 Stori...   
1895        1896   278390                                   CHILD IS BORN, A   
1896        1897   278390                                 Business in Action   
1897        1898   278390                               Stimulus Factor, The   

      Book.Rating  
0               5  
1886            8  
1887            7  
1888            7  
1889            7  
1890            4  
1891            7  
1892            8  
1893            9  
1894            7  
1895           10  
1896            7  
1897            9

In [26]:
book_data[(book_data['User_ID']==276736) | (book_data['User_ID']==276726)]

Unnamed: 0  User_ID                                         Book.Title  \
0           1   276726                                Classical Mythology   
3           4   276736  Flu: The Story of the Great Influenza Pandemic...   

   Book.Rating  
0            5  
3            8

In [30]:
def get_recommendation(customer_id):
    tem=list(cosine_df.sort_values([customer_id],ascending=False).head(3).index)
    book_list=list(book_data[book_data['User_ID'].isin(tem)]['Book.Title'])
    recommended_books=set(book_list)-set(book_data[book_data['User_ID']==customer_id]['Book.Title'])
    return recommended_books

In [31]:
get_recommendation(276729)

{'A Man of Steel',
 'Classical Mythology',
 'The Apprentice',
 'The Seventh Night'}